### ISSUE TITLE
Performance Regression in v5.1.20-prerelease

### ISSUE THREAD

**Issue by Jermolene** [Jermolene/TiddlyWiki5#3998]
@bimlas reports [over on the mailing list](https://groups.google.com/d/msgid/tiddlywikidev/9f1ddb56-9ca6-4133-9e64-c8625036ae32%40googlegroups.com) that there has been a significant regression in the performance of the [kin filter](https://github.com/bimlas/tw5-kin-filter/blob/66e2cbbf85761daa4c33e9c422db0fce3b6ea6f9/plugins/kin-filter/kin.js).

The steps to repeat the tests are:

* Download current and prerelease
* Install Kin filter from https://bimlas.gitlab.io/tw5-kin-filter/
* Enable performance instrumentation
* Save and reload wikis
* Close all opened tiddlers and open "Open" sidebar
* Open Developer Tools Console
* Use "[kin::to[TableOfContents]]" filter in Advanced Search

The timings from @bimlas were as follows:

* Chrome 75.0.3770.90
  * Current:
    * mainRefresh: 861.93ms 
    * kin filter: n/a
  * Prerelease
    * mainRefresh: 1317.06ms 
    * kin filter: invocations 3; avgTime 331ms; totalTime 994ms; percentTime 78ms
  * Prerelease (indexing disabled)
    * mainRefresh: 1206ms/1285ms
    * kin filter: invocations 3; avgTime 301ms/532ms; totalTime 902ms; percentTime 82ms
* Firefox 67.0.2
  * Current
    * mainRefresh: 890.00ms
    * +filter: 657.00ms 
  * Prerelease
    * mainRefresh: 1108.00ms 
    * kin filter: invocations 3; avgTime 270ms; totalTime 810ms; percentTime 72ms
  * Prerelease (indexing disabled)
    * mainRefresh: 1141ms/1078ms
    * kin filter: invocations 3; avgTime 275ms/377ms; totalTime 825ms; percentTime 47.8ms

These measurements strongly imply that v5.1.20 has introduced an overall performance regression. However, there are some inconsistencies -- for example, the double timings for the prerelease with indexing disabled were the result of repeating the tests.

To aid further investigation it is proposed to build a test rig for more accurately measuring performance regressions: to be able to run a performance test a few thousand times across each commit in a range and tabulate the results.

**Comment by bimlas** [comment]
I'm very sorry, but it was a false alarm.

I tried to find out the cause of the issue with Git bisect, and as I moved forward in history, I found the Kin filter slower. There was no sudden change in speed, but it slowed with every commit. So it seemed that the "error" was constantly increasing.

One essential thing I failed to recognize was the fact that in the test I looked at the time it takes to find all successors of TableOfContents - but I didn't noticed that in prerelease are more tiddlers connecting to ToC (1002 tiddlers) than in 5.1.19 (932 tiddlers). The discovery of more tiddlers takes more time, so Tiddly did not get slower, only the filter has to find more tiddlers.

To check the correctness of the theory, I imported the documentation tiddlers to 5.1.19 that are included in the prerelease only (`git diff --name-only --diff-filter=A v5.1.19 -- editions/`) and when I ran the test, I've got the same results as on prerelease.
